client/src/main/java/org/apache/celeborn/client/ReviveManager.java

```diff
@@ -0,0 +1,125 @@
+                  Set<ReviveRequest> requests = shuffleEntry.getValue();
+                  Set<Integer> mapIds = new HashSet<>();
+                  ArrayList<ReviveRequest> filteredRequests = new ArrayList<>();
+                  Map<Integer, ReviveRequest> requestsToSend = new HashMap<>();
+
+                  // Insert request that is not MapperEnded and with the max epoch
+                  // into requestsToSend
+                  Iterator<ReviveRequest> iter = requests.iterator();
+                  while (iter.hasNext()) {
+                    ReviveRequest req = iter.next();
+                    if (shuffleClient.checkRevivedLocation(
+                        shuffleId, req.mapId, req.partitionId, req.loc.getEpoch(), false)) {
+                      req.reviveStatus = StatusCode.SUCCESS;
+                    } else {
+                      filteredRequests.add(req);
+                      mapIds.add(req.mapId);
+                      PartitionLocation loc = req.loc;
+                      if (!requestsToSend.containsKey(loc.getId())
+                          || requestsToSend.get(loc.getId()).epoch < loc.getEpoch()) {
+                        requestsToSend.put(loc.getId(), req);
+                      }
+                    }
+                  }
+
+                  if (!requestsToSend.isEmpty()) {
+                    // Call reviveBatch. Return null means Exception caught or
+                    // SHUFFLE_NOT_REGISTERED
+                    Map<Integer, StatusCode> results =
+                        shuffleClient.reviveBatch(
+                            appId, shuffleId, mapIds, requestsToSend.values());
+                    if (results == null) {
+                      for (ReviveRequest req : filteredRequests) {
+                        req.reviveStatus = StatusCode.REVIVE_FAILED;
+                      }
+                    } else {
+                      for (ReviveRequest req : filteredRequests) {
+                        if (shuffleClient.mapperEnded(shuffleId, req.mapId)) {
+                          req.reviveStatus = StatusCode.MAP_ENDED;
+                        } else {
+                          req.reviveStatus = results.get(req.partitionId);
+                        }
+                      }
+                    }
+                  }
+                } // End foreach shuffleMap.entrySet
+              } // End foreach inProcessRequests.entrySet
+            }
+
+            inProcessRequests.clear();
+          }
+        },
+        100,
+        100,
+        TimeUnit.MILLISECONDS);
+  }
+
+  public void addRequest(String appId, int shuffleId, ReviveRequest request) {
+    shuffleClient.excludeWorkerByCause(request.cause, request.loc);
+    // This sync is necessary to ensure the add action is atomic
+    synchronized (this) {
+      HashMap<Integer, Set<ReviveRequest>> shuffleMap =
+          pendingRequests.computeIfAbsent(appId, (id) -> new HashMap());
```

Comment:
Why need the app id? It work's only for one app

Reply:
I am also a little confused about this. If needs to support spark multi-contexts, like the old version of kyuubi w/ spark2, then ShuffleClient may also need to add appid to other data structures related to shuffleId. If it is not supported or not allowed in Celeborn, some places where appId is involved would be simplified, such as here. 
Another point is if the ContextCleaner can clean up rss state involved when it stops old context in spark3. Only a rough idea, welcome to correct me.

Reply:
what's the flink case? @RexXiong

Reply:
It's same, only one appId active.

Reply:
done